### ISSUE TITLE
JSON pretty print collapses whitespace in between brackets (causing random test failure)

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#1940]
We recently had a test failure with this message:

```
---- User: User email administration (UserAdminSettingsEmailFormTest) ----
Status    Group      Filename          Line Function                            
--------------------------------------------------------------------------------

Fail      Other      user.test         1519 SystemConfigFormCase->testConfigFor
    Value ".Q[ ]Uz@" is equal to value ".Q[]Uz@".
```

This indicates that spaces in square brackets are getting collapsed incorrectly. This was intended to fix differences between versions of PHP when dealing with empty arrays, but it looks like it's collapsing spaces in ways that we had not intended.

Related to the meta issue for fixing random failures at #1478.

**Comment by Gormartsen** [comment]
@quicksketch 
I was doing pretty json encode as well for my own project here:
https://github.com/itpatrol/drupal_deploy/blob/7.x/drupal_deploy.module#L121

This code support UTF8 strings . like  `value : "русский"` instead of `value: "\u0440\u0443\u0441\u0441\u043A\u0438\u0439"`

Maybe you can adopt some part of the code to support pretty json for php before 5.4.0

**Comment by quicksketch** [comment]
> Maybe you can adopt some part of the code to support pretty json for php before 5.4.0

One of the weird things about this issue is that the collapsing of brackets actually needs to be run on some versions of PHP 5.4 as well (up to and including PHP 5.4.27). I think most other differences between PHP versions is already handled, as are UTF-8 strings. We could probably still adopt some of the code if it turns out to be more performant or easier to read.

**Comment by Gormartsen** [comment]
Yeah. It's quite mess there. 

Most visible part is that there is a recursion that allow keep pretty format to multilevel objects or arrays. Other code oriented on proper encoding values and escape changes to specific characters to do not brake Json format.

Regards,
Gor Martsen
@gormartsen

**Comment by quicksketch** [comment]
The additional tests in my PR at https://github.com/backdrop/backdrop/pull/1422 turned up a different bug in the pretty-printing as well. It turns out the unescaping of Unicode characters was unescaping control characters like NULL, backspace, end-of-line, etc, which results in invalid JSON that can't be decoded. I updated the PR to fix this issue and tests should now be passing.

**Comment by quicksketch** [comment]
Merged in the fix for 1.4.3.
